nifpga/session.py

```diff
@@ -456,226 +449,214 @@ def __init__(self,
+        For example, if the register had a word length of 54, the 54 MSB would
+        be the fixed point bits. The 10 LSB of the combinedData must be shifted
+        off in order to not mess up further calculations.
+
+        """
         combinedData = 0
-        for index in range(0, len(self)):
-            combinedData = (combinedData << 32 * index) + data[index]
-        if len(self) > 1:
+        for index in range(0, self._transfer_len):
+            combinedData = (combinedData << 32) + data[index]
+        if self._transfer_len > 1:
             overflow_bit = 1 if self._overflow_enabled else 0
             total_num_bits = self._word_length + overflow_bit
-            combinedData = combinedData >> (32 * len(self) - total_num_bits)
+            combinedData = combinedData >> (32 * self._transfer_len - total_num_bits)
         return combinedData
 
     def _convert_from_read_value_to_decimal(self, data):
-        """ This function converts read value from a RIO device and sets
-        the overflow bit if enabled and returns a Decimal representation of
-        the fixed point. """
+        """ This method converts value from hardware and returns the respective
+        decimal value or a tuple with the overflow status and the decimal
+        value.
+        """
+        overflow = None
         if self._overflow_enabled:
-            self.overflow = self._get_overflow_value(data)
+            overflow = self._get_overflow_value(data)
             data = self._remove_overflow_bit(data)
+
         if self._signed:
             data = self._integer_twos_comp(data)
-        return Decimal(data * self._delta)
+        decimal_value = Decimal(data * self._delta)
+        if self._overflow_enabled:
+            return (overflow, decimal_value)
+        else:
+            return decimal_value
 
     def _get_overflow_value(self, data):
         """ Mask out all the data within the word length, leaving the overflow
         bit. If the result after masking the the word portion of the fixed
-        point is none zero that indicates the data read has overflown."""
+        point is none zero that indicates the data read has overflowed. """
         mask = 2**(self._word_length)
         if data & mask > 0:
             return True
         return False
 
     def _remove_overflow_bit(self, data):
-        """ This helper function masks out all bits not inside the word length,
-        ultimately returning a value of data missing the overflow bit. """
+        """ This helper method masks out all bits not inside the word length,
+        ultimately returning a value of data without the overflow bit. """
         return data & (2**(self._word_length) - 1)
 
     def _integer_twos_comp(self, data):
         if (data & (1 << (self._word_length - 1))) != 0:
-            data = data - (1 << self._word_length)
+            data = data ^ (2 ** (self._word_length) - 1)
+            data += 1
+            data *= -1
         return data
```

Comment:
The previous version was a one-liner: "data - (1 << self._word_length)". Is the new version somehow better? I don't find it any more intuitive, and I'd ask for a comment or two to explain what's going on.